### PR TITLE
Prevent FIN_WAIT connections by using stream.destorySoon()

### PR DESCRIPTION
**Pull request by eladnava**
As per #306.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -608,7 +608,13 @@ Client.prototype.close = function(callback, reason) {
     // needed in case of errors
     if (!that._closed) {
       cleanup();
-      that.connection.stream.end();
+      // prefer destroySoon() as it prevents FIN_WAIT connections
+      if (that.connection.stream.destroySoon) {
+        that.connection.stream.destroySoon();
+      }
+      else {
```
Comment (mcollina) [inline lib/client.js]: can you please move the else inline with the `}`?

**Comment by mcollina** [comment]
I'm thinking we can have a three way case: `destroySoon`, `destroy` and `end`. In this order.

What do you think?

**Comment by eladnava** [comment]
@mcollina Done, however I wasn't able to run the tests completely as I don't have Redis installed.

**Comment by mcollina** [comment]
@behard what do you think?

**Comment by eladnava** [comment]
I think you meant @behrad =)

**Comment by mcollina** [comment]
Oh yes :D.

**Comment by behrad** [comment]
So it should reduce `FIN_WAIT`s :)
If destroySoon is preferable, how will that affect on the mobile/public facing client end?

**Comment by eladnava** [comment]
@behrad When connected clients passively disconnect from the internet, they will eventually be kicked due to Mosca's [keepalive timeout mechanism](https://github.com/mcollina/mosca/blob/05083a49f61c246c0db85c3c8815dda2fede5e9e/lib/client.js#L171), which calls `socket.end()` if the client hasn't sent an MQTT keep alive packet within the client-defined time limit.

In its current implementation, calling `socket.end()` on those connections will have them stuck (if not indefinitely, then for a long time) in FIN_WAIT states because the TCP stack will attempt to let the client know it's closing down the connection. But since the client is no longer reachable then the server will never be able to deliver the FIN/ACK packet to the client, nor will it receive a FIN/ACK or an ACK from the client in response to the FIN/ACK.

ssldig09

Does that answer your question?

**Comment by behrad** [comment]
great explanation @eladnava 👍 
I can test to see how this patch will change the real world scenario :)

**Comment by eladnava** [comment]
@behrad excellent, please let us know if it's OK to merge :)

**Comment by mcollina** [comment]
Ping @behrad? I'll like to merge this and release next week!

**Comment by behrad** [comment]
Sorry guys, so busy these days, it takes me some time to test this, I'l let you even after @mcollina releases this :)

**Comment by eladnava** [comment]
@mcollina Has this been released yet? I'd love to use it in production. Thanks! 😄

**Comment by mcollina** [comment]
Sorry, released now as v2.1.0.

**Comment by eladnava** [comment]
Excellent, thank you! 👍

**Pull request merged**